packages/liveblocks-node/src/client.ts

```diff
@@ -299,21 +878,26 @@ export class Liveblocks {
   }): Promise<CommentData> {
     const { roomId, threadId, commentId } = params;
 
-    const resp = await this.get(
+    const res = await this.get(
       `/v2/rooms/${encodeURIComponent(roomId)}/threads/${encodeURIComponent(
         threadId
       )}/comments/${encodeURIComponent(commentId)}`
     );
 
-    const body = await (resp.json() as Promise<CommentData>);
-
-    if (resp.status !== 200) {
-      throw {
-        status: resp.status,
-        ...body,
-      };
+    if (!res.ok) {
+      const text = await res.text();
+      throw new ApiError(res.status, text);
     }
 
-    return body;
+    return res.json() as Promise<CommentData>;
+  }
+}
+
+export class ApiError extends Error {
```

Comment:
I saw this [comment](https://github.com/liveblocks/liveblocks/pull/1280#discussion_r1379827639) about custom errors on a different PR. So I'm preemptively going to give my opinion on this just in case @nvie jumps in :D 

Like @nvie, I'm also not a fan of extending `Error` to add extra data to an error. However, it seems to be the most common and recommended pattern in the JS ecosystem. So I like what you did here. I agree with exporting a custom error, but maybe `HttpError` instead? Usually, I also add a `type` field so the consumer can use a switch and not have to use `instanceOf`.

We can keep it as `@internal` until we have a plan to expose custom errors at large. My only requirement for that plan is that we don't fight the ecosystem/language for purity sake (e.g. force people to handle errors to read the data).

Reply:
+1 to changing the name to `HttpError`, because… that's what it seems to be.

Adding a `type` field won't work with error handling, because the only reasonable way to handle errors in TypeScript is by catching a value as `unknown`, i.e.

```ts
try {
  ...
} catch (e: unknown) {
  if (e instanceof HttpApi) { ... }     // alternative 1
  if (e.type === 'HttpError') { ... }   // alternative 2
  if (isHttpError(e)) { ... }           // alternative 3
  if (client.isError(e)) { ... }        // alternative 4
}
```

Therefore, alternative 1 works, but it means we'll have to expose the HttpError class as an implementation detail and we're stuck with maintaining it over time.

Alternative 2 that @GuillaumeSalles suggested won't be nice DX in TypeScript, because `e` is `unknown`, and TypeScript won't let you read the `.type` field off of it.

I'd personally be in favor of using alternative 3 or 4. In alternative 3, we could exposing a helper function to be imported `isHttpError()` (or similarly named). Alternative 4 is basically the same, but by exposing it on the `client` that's already in scope for the user, they can avoid the extra import.

```ts
export function isHttpError(e: unknown): e is HttpError {
  return e instanceof HttpError;
}
```

Using this function will let TypeScript refine the value, so you can access members afterwards:

```ts
try {
  ..
} catch (e: unknown) {
  if (client.isError(e)) {
    e.status  // ✅ number
    e.message  // ✅ string
  }
}
```

What do you think?

Reply:
What are the benefits of exposing a function that does `instanceof` instead of the user doing `instanceof` directly?

Doing `instanceof` for customer error handling is the most common and recommended pattern. Expected by developers of all levels.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#handling_a_specific_error_type

Reply:
Two benefits:

- We'd not have to commit to the name `HttpError` just yet and can keep it an implementation detail. If we want to change the custom error name later, it's not a breaking change. (This doesn't hold for the _fields_ in the error of course.)
- Users don't have to import a custom error type just to be able to handle it. Reason I suggested using putting a method on the client is that that variable is already in scope, no extra import needed.

To me, it feels quite wrong to be exporting a non-standard `HttpError` class from a (Liveblocks-specific) library, mostly. If all library authors did this, then as a user you'd be importing and handling a ton of different "HttpError"s from different libraries 😬

Reply:
I mean, I could live with alternative 1 if we think it's more intuitive. In a way, this suggestion is a way to punt on the decision of how we want to do error handling until we have an API we are happy with and willing to commit to in the long term.

Reply:
> We'd not have to commit to the name HttpError just yet and can keep it an implementation detail. If we want to change the custom error name later, it's not a breaking change. (This doesn't hold for the fields in the error of course.)

If we do this, we're committing to a function name instead of class name, so I don't think this argument is valid.

> Users don't have to import a custom error type just to be able to handle it. Reason I suggested using putting a method on the client is that that variable is already in scope, no extra import needed.

It's more important to have an expected pattern than avoid an import. 

- Stripe is exporting `StripeError`s for [error handling](https://github.com/stripe/stripe-node/wiki/Error-Handling#typescript)

- Slack is exporting [custom errors](https://www.npmjs.com/package/@slack/web-api?activeTab=code) under `dist/errors.d.ts`

- Resend is exporting [ResendError](https://www.npmjs.com/package/resend?activeTab=code) under `build/src/error.d.ts`

I'm not using a lot of REST client but this is the pattern I always encountered/expected since I used typescript. Happy to look into other REST API wrappers if you think I'm currently biased.

(They all also include a code to let the user do a `switch` after checking the type with `instanceof`. That's what I meant in my original message.)

If I see a `StripeError` in my console logs, without looking at the docs, my reflex is to try to import `StripeError` and not look for method on the client to check the type of the error.

> To me, it feels quite wrong to be exporting a non-standard HttpError class from a (Liveblocks-specific) library, mostly. If all library authors did this, then as a user you'd be importing and handling a ton of different "HttpError"s from different libraries 😬

That's why I went for `LiveblocksError` in the client (following Stripe pattern at the time) to avoid any collision (even if I never exported it).

Reply:
A'ight, you convinced me. Let's go with `LiveblocksError` in that case 👍